src/modules/rlm_mruby/rlm_mruby.c

```diff
@@ -0,0 +1,468 @@
+			const char *key = cf_pair_attr(cp);
+			const char *value = cf_pair_value(cp);
+			mrb_value mrubyKey, mrubyValue;
+
+			if (!key || !value) continue;
+
+			mrubyKey = mrb_str_new_cstr(mrb, key);
+			mrubyValue = mrb_str_new_cstr(mrb, value);
+
+			if (!mrb_nil_p(mrb_hash_get(mrb, hash, mrubyKey))) {
+				WARN("rlm_mruby: Ignoring duplicate config item '%s'", key);
+				continue;
+			}
+
+			mrb_hash_set(mrb, hash, mrubyKey, mrubyValue);
+
+			DEBUG("%*s%s = %s", indent_item, " ", key, value);
+		}
+	}
+
+	DEBUG("%*s}", indent_section, " ");
+}
+
+/*
+ *	Do any per-module initialization that is separate to each
+ *	configured instance of the module.  e.g. set up connections
+ *	to external databases, read configuration files, set up
+ *	dictionary entries, etc.
+ */
+static int mod_instantiate(UNUSED CONF_SECTION *conf, void *instance)
+{
+	rlm_mruby_t *inst = instance;
+	mrb_state *mrb;
+	CONF_SECTION *cs;
+	FILE *f;
+	mrb_value status;
+
+	mrb = inst->mrb = mrb_open();
+	if (!mrb) {
+		ERROR("mruby initialization failed");
+		return -1;
+	}
+
+	/* Define the freeradius module */
+	DEBUG("Creating module %s", inst->module_name);
+	inst->mruby_module = mrb_define_module(mrb, inst->module_name);
+	if (!inst->mruby_module) {
+		ERROR("Creating module %s failed", inst->module_name);
+		return -1;
+	}
+
+	/* Define the radlog method */
+	mrb_define_class_method(mrb, inst->mruby_module, "radlog", mruby_radlog, MRB_ARGS_REQ(2));
+
+#define A(x) mrb_define_const(mrb, inst->mruby_module, #x, mrb_fixnum_value(x));
+	/* Define the logging constants */
+	A(L_DBG);
+	A(L_WARN);
+	A(L_AUTH);
+	A(L_INFO);
+	A(L_ERR);
+	A(L_PROXY);
+	A(L_WARN);
+	A(L_ACCT);
+	A(L_DBG_WARN);
+	A(L_DBG_ERR);
+	A(L_DBG_WARN_REQ);
+	A(L_DBG_ERR_REQ);
+
+	/* Define the return value constants */
+	A(RLM_MODULE_REJECT)
+	A(RLM_MODULE_FAIL)
+	A(RLM_MODULE_OK)
+	A(RLM_MODULE_HANDLED)
+	A(RLM_MODULE_INVALID)
+	A(RLM_MODULE_USERLOCK)
+	A(RLM_MODULE_NOTFOUND)
+	A(RLM_MODULE_NOOP)
+	A(RLM_MODULE_UPDATED)
+	A(RLM_MODULE_NUMCODES)
+#undef A
+
+	/* Convert a FreeRADIUS config structure into a mruby hash */
+	inst->mrubyconf_hash = mrb_hash_new(mrb);
+	cs = cf_section_sub_find(conf, "config");
+	if (cs) mruby_parse_config(mrb, cs, 0, inst->mrubyconf_hash);
+
+	/* Define the Request class */
+	inst->mruby_request = mruby_request_class(mrb, inst->mruby_module);
+
+	DEBUG("Loading file %s...", inst->filename);
+	f = fopen(inst->filename, "r");
+	if (!f) {
+		ERROR("Opening file failed");
+		return -1;
+	}
+
+	status = mrb_load_file(mrb, f);
+	if (mrb_undef_p(status)) {
+		ERROR("Parsing file failed");
+		return -1;
+	}
+	fclose(f);
+
+	status = mrb_funcall(mrb, mrb_obj_value(inst->mruby_module), "instantiate", 0);
+	if (mrb_undef_p(status)) {
+		ERROR("Running instantiate failed");
+		return -1;
+	}
+
+	return 0;
+}
+
+#define BUF_SIZE 1024
+static mrb_value mruby_vps_to_ary(mrb_state *mrb, VALUE_PAIR **vps)
+{
+	mrb_value res;
+	VALUE_PAIR *vp;
+	vp_cursor_t cursor;
```

Comment:
That should be fr_cursor_t, yes I know it's confusing hopefully we'll hopefully unify them again soon.